### DISCUSSION TITLE
Tuya Water valve TS0049 made a start but need some help

### DISCUSSION THREAD

**Discussion by toogooda** [Koenkk/zigbee2mqtt#16377] [Devices]
Raised the following issue : [https://github.com/Koenkk/zigbee2mqtt/issues/15124](https://github.com/Koenkk/zigbee2mqtt/issues/15124)

I have created a converter and can now see changes when i turn on and off but cannot remotely turn on and off. Also 'get' 'state' fails if i click the refresh.

I have all the tuya data-points for it.
```

const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const utils = require('zigbee-herdsman-converters/lib/utils');
const e = exposes.presets;
const ea = exposes.access;
const tuya = require('zigbee-herdsman-converters/lib/tuya');

const definition = {
    // Since a lot of TuYa devices use the same modelID, but use different datapoints
    // it's necessary to provide a fingerprint instead of a zigbeeModel
    fingerprint: [
        {
            // The model ID from: Device with modelID 'TS0601' is not supported
            // You may need to add \u0000 at the end of the name in some cases
            modelID: 'TS0049',
            // The manufacturer name from: Device with modelID 'TS0601' is not supported.
            manufacturerName: '_TZ3210_0jxeoadc',
        },
    ],
    model: 'Rain Seer',
    vendor: 'TuYa Compatable',
    description: 'Zigbee Hose Controller',
    fromZigbee: [tuya.fz.datapoints],
    toZigbee: [tuya.tz.datapoints, tz.tuya_switch_state],
    onEvent: tuya.onEventSetTime, // Add this if you are getting no converter for 'commandMcuSyncTime'
    configure: tuya.configureMagicPacket,
    exposes: [e.switch() , e.battery(), tuya.exposes.batteryState(),e.temperature()
        // Here you should put all functionality that your device exposes
    ],
    meta: {
        // All datapoints go in here
        tuyaDatapoints: [
            [101, 'state', tuya.valueConverter.onOff], // <- mapped the datapoint
            [120, 'battery_state', tuya.valueConverter.batteryState],
            [115, 'battery', tuya.valueConverter.raw],
            [113,'temperature', tuya.valueConverter.raw]

        ],
    },
};

module.exports = definition;
``` 

Can anyone help?

**Comment by tzeryin1981** [comment]
I think i have the same device with you, same issue here

**Reply by DarioKTS** [reply]
I've just purchased a bunch of the same valves!
Totally new at writing converters for Zigbee device.
Will use the code above as a starting point and report back :)

**Reply by toogooda** [reply]
Not a lot working so far sorry, if you change the state i.e. if you turn it on or off manually then it is reflected correctly in Zigbee2MQTT but you cannot change it in Z2M so not very useful :(
Have been trying to get some help but no one so far.
SOme battery stuff works also but takes a day or so to come through.
I have ordered a tuya hub so i can see all the data points etc and hoping it will help me get it going.
Please jump in the linked raised issue and upvote it :)

**Reply by tzeryin1981** [reply]
There is no code reference for TS0049 in the Herdman list, are we miss out something?

**Reply by DarioKTS** [reply]
I think i'm in need of direction...
I may have another version of the same valve OR the same rebranded.. 

My devices are recognised as 'TS0049', the manufacturer however is '_TZ3000_hwnphliv' (FrankEver model FK-WT03W, zigbee version, Tuya compatible, not yellow green but all black, physically otherwise seems the same to the rainseer one i thnk).
<img width="598" alt="image" src="https://user-images.githubusercontent.com/118348582/217194374-f042290e-bf19-4bba-bd5e-728183643aa2.png">

So here's the question for those of you who are mnore experienced: 
Should i open a new topic + issue, OR is thisthe same valve rebranded?

**Reply by tzeryin1981** [reply]
believe is the same tuya module, if we figure what is the root of cause, then it solve our problem.

**Reply by DarioKTS** [reply]
anyhow, i have an update, see if it might be useful to you :)

with htis converter i get both remote and local control over switching on and off the valve + over a number of behaviours of the device, battery reporting is jumpy (not sure why, once is 29% once 100%) but .. at least it shows up.
<img width="514" alt="image" src="https://user-images.githubusercontent.com/118348582/217274216-3711b4e5-4b91-4381-9799-ba17ae05fc1d.png">


I still haven't gotten around the tuya datapoints, i was trying to solve the "inclusions" (sorry if it isn't the correct term, i mean the -fromZigbee:[] and -toZigbee[] lines in the converter and having them mirror the -exposes line

```
const fz = {...require('zigbee-herdsman-converters/converters/fromZigbee'), legacy: require('zigbee-herdsman-converters/lib/legacy').fromZigbee};
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const tuya = require('zigbee-herdsman-converters/lib/tuya');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;

module.exports = {
    fingerprint: [{modelID: 'TS0049', manufacturerName: '_TZ3000_hwnphliv'}],
    model: 'FK-WT03W',
    vendor: 'Frankever',
    description: 'Smart irrigation valve',
    fromZigbee: [fz.battery, tuya.fz.power_on_behavior, tuya.fz.switch_type, fz.on_off, ],
    toZigbee: [tuya.tz.power_on_behavior, tuya.tz.switch_type, tz.on_off ],
    onEvent: tuya.onEventSetLocalTime,
    configure: tuya.configureMagicPacket,
    exposes: [ e.power_on_behavior(), tuya.exposes.switchType(), e.switch(), e.battery(), e.battery_voltage() ],
    meta: {
        tuyaDatapoints: [
        ],
    },
};
```


When first connected , withempty converter, i had a log like this
`Received Zigbee message from 'VALVE', type 'attributeReport', cluster 'manuSpecificTuya_3', data '{"powerOnBehavior":0,"switchType":0}' from endpoint 1 with groupID 0`
this led me to tuya.tz.power_on_behavior and e.power_on_behavior()


there are few more logs that i havne't gotten to solve yet and Z2M complans that has no no converter for, like:

`Received Zigbee message from 'VALVE', type 'attributeReport', cluster 'genOnOff', data '{"32768":0,"moesStartUpOnOff":0,"offWaitTime":0,"onOff":0,"onTime":0,"tuyaBacklightMode":1,"tuyaBacklightSwitch":0}' from endpoint 1 with groupID 0`

`Received Zigbee message from 'VALVE', type 'attributeReport', cluster 'genBasic', data '{"65503":"\u0000\u0000\u0000\u0000\u0011"}' from endpoint 1 with groupID 0`

`Received Zigbee message from 'VALVE', type 'read', cluster 'genTime', data '["localTime"]' from endpoint 1 with groupID 0`

ans a few array like messages

`Received Zigbee message from 'VALVE', type 'raw', cluster 'manuSpecificTuya_3', data '{"data":[8,81,10,64,208,72,6,11,0,0,2,28],"type":"Buffer"}' from endpoint 1 with groupID 0`

`Received Zigbee message from 'VALVE', type 'raw', cluster 'manuSpecificTuya_3', data '{"data":[8,95,10,64,208,72,3,10,0],"type":"Buffer"}' from endpoint 1 with groupID 0`

Exploring the tuya converters, the devices with e.power_on_behaviour() tend to have a bit of code instead of the datapoints, like "configure: async(...)
await tuya.configureMagicPacket
const endpoint= ...
await reporting ..."

I don't feel confident tackling that as of yet ...maybe someone experienced can jump in and solve it all with ease?

**Reply by tzeryin1981** [reply]
Are you mean that datapoints coding does not work with this?

**Reply by DarioKTS** [reply]
more like:
I wasn't getting anywhere with the datapoints, i'm sure someone experienced might.
.... But i got something to work, so see if it helps you :)

ps - rememeber to match the manufacturer to your own if you do try

**Reply by tzeryin1981** [reply]
No luck, tried, still no working.....

**Reply by DarioKTS** [reply]
edit of the previous, might help, still testing

```
const fz = {...require('zigbee-herdsman-converters/converters/fromZigbee'), legacy: require('zigbee-herdsman-converters/lib/legacy').fromZigbee};
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const tuya = require('zigbee-herdsman-converters/lib/tuya');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;

module.exports = {
    fingerprint: [{modelID: 'TS0049', manufacturerName: '_TZ3000_hwnphliv'}],
    model: 'FK-WT03W',
    vendor: 'Frankever',
    description: 'Smart irrigation valve',
    fromZigbee: [fz.battery, tuya.fz.power_on_behavior, tuya.fz.switch_type, fz.on_off, ],
    toZigbee: [tuya.tz.power_on_behavior, tuya.tz.switch_type, tz.on_off ],
    onEvent: tuya.onEventSetLocalTime,
    configure: tuya.configureMagicPacket,
    exposes: [ e.power_on_behavior(), tuya.exposes.switchType(), e.switch(), e.battery(), e.battery_voltage() ], 
  meta: {
           configure: async (device, coordinatorEndpoint, logger) => {
            try {
                const endpoint = device.getEndpoint(1);
                await reporting.bind(endpoint, coordinatorEndpoint, ['genPowerCfg']);
            } catch (error) {/* Fails for some*/}
            },
        tuyaDatapoints: [
        ],
    },
};
```
the last part binds endpoint1, i might have lost it in between revisions.

I'm getting a tuya hub, this thing is driving me nuts and i have to finish asap.

**Reply by toogooda** [reply]
I can confirm that your device is not the same, i wasn't able to reproduce any of your success so my guess is that you have a version with at least different data points.

Looking at your logs I would have a look at [these examples](https://github.com/Koenkk/zigbee-herdsman-converters/blob/master/devices/tuya.js) esp this one for clues as it makes a ref to "moesStartUpOnOff" which was listed in your logs

```
 description: 'Smart water/gas valve',
        extend: tuya.extend.switch({powerOnBehavior: true}),
        configure: async (device, coordinatorEndpoint, logger) => {
            await tuya.configureMagicPacket(device, coordinatorEndpoint, logger);
            const endpoint = device.getEndpoint(1);
            await endpoint.read('genOnOff', ['onOff', 'moesStartUpOnOff']);
        },
```
I suggest you create a new issue as people with Yellow Rain Seer will not be able to use your code.

Good luck

**Comment by tzeryin1981** [comment]
You need the datapoint code? I can list down to you.

**Comment by SteveShirvo** [comment]
> Raised the following issue : #15124
> 
> I have created a converter and can now see changes when i turn on and off but cannot remotely turn on and off. Also 'get' 'state' fails if i click the refresh.
> 
> I have all the tuya data-points for it.
> 
> ```
> 
> const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
> const tz = require('zigbee-herdsman-converters/converters/toZigbee');
> const exposes = require('zigbee-herdsman-converters/lib/exposes');
> const reporting = require('zigbee-herdsman-converters/lib/reporting');
> const extend = require('zigbee-herdsman-converters/lib/extend');
> const utils = require('zigbee-herdsman-converters/lib/utils');
> const e = exposes.presets;
> const ea = exposes.access;
> const tuya = require('zigbee-herdsman-converters/lib/tuya');
> 
> const definition = {
>     // Since a lot of TuYa devices use the same modelID, but use different datapoints
>     // it's necessary to provide a fingerprint instead of a zigbeeModel
>     fingerprint: [
>         {
>             // The model ID from: Device with modelID 'TS0601' is not supported
>             // You may need to add \u0000 at the end of the name in some cases
>             modelID: 'TS0049',
>             // The manufacturer name from: Device with modelID 'TS0601' is not supported.
>             manufacturerName: '_TZ3210_0jxeoadc',
>         },
>     ],
>     model: 'Rain Seer',
>     vendor: 'TuYa Compatable',
>     description: 'Zigbee Hose Controller',
>     fromZigbee: [tuya.fz.datapoints],
>     toZigbee: [tuya.tz.datapoints, tz.tuya_switch_state],
>     onEvent: tuya.onEventSetTime, // Add this if you are getting no converter for 'commandMcuSyncTime'
>     configure: tuya.configureMagicPacket,
>     exposes: [e.switch() , e.battery(), tuya.exposes.batteryState(),e.temperature()
>         // Here you should put all functionality that your device exposes
>     ],
>     meta: {
>         // All datapoints go in here
>         tuyaDatapoints: [
>             [101, 'state', tuya.valueConverter.onOff], // <- mapped the datapoint
>             [120, 'battery_state', tuya.valueConverter.batteryState],
>             [115, 'battery', tuya.valueConverter.raw],
>             [113,'temperature', tuya.valueConverter.raw]
> 
>         ],
>     },
> };
> 
> module.exports = definition;
> ```
> 
> Can anyone help?

**Comment by SteveShirvo** [comment]
Hi have you had any joy with this yet, I've been struggling for months to find an workable process.
